ujit_codegen.c

```diff
@@ -935,6 +935,67 @@ gen_opt_plus(jitstate_t* jit, ctx_t* ctx)
     return true;
 }
 
+static bool
+gen_opt_nil_p(jitstate_t* jit, ctx_t* ctx)
+{
+    // If someone redefined nil? on NilClass, we'll never JIT this instruction
+    if (!BASIC_OP_UNREDEFINED_P(BOP_NIL_P, NIL_REDEFINED_OP_FLAG)) {
+        return false;
+    }
+
+    // Create a size-exit to fall back to the interpreter
+    // Note: we generate the side-exit before popping operands from the stack
+    uint8_t* side_exit = ujit_side_exit(jit, ctx);
+
+    struct rb_call_data * cd = (struct rb_call_data *)jit_get_arg(jit, 0);
+
+    assume_method_lookup_stable(cd->cc, vm_cc_cme(cd->cc), jit->block);
+
+    x86opnd_t arg0 = ctx_stack_pop(ctx, 1);
+
+    cmp(cb, arg0, imm_opnd(Qnil));
+    int NOT_Qnil = cb_new_label(cb, "NOT_Qnil");
+    int DONE = cb_new_label(cb, "DONE");
+
+    jne(cb, NOT_Qnil);
+
+    // If the operand is Qnil, check that nil? hasn't been redefined
+    mov(cb, RAX, const_ptr_opnd(ruby_current_vm_ptr));
+    test(
+        cb,
+        member_opnd_idx(RAX, rb_vm_t, redefined_flag, BOP_NIL_P),
+        imm_opnd(NIL_REDEFINED_OP_FLAG)
+    );
+    // If it has been redefined, fall back to the interpreter
+    jnz_ptr(cb, side_exit);
+
+    // nil? confirmed
+    x86opnd_t dst = ctx_stack_push(ctx, T_TRUE);
```

Comment:
I don't think you should put `T_TRUE` here if you can write `Qfalse` on the stack later. Use `T_NONE` instead.

Reply:
Thanks, I fixed it.  I wasn't sure what the parameter was for 😅